File: src/core/vercel-actions.ts
import {
  normalizeEnvVar,
  vercelRequest,
  type RawEnvVar,
  type VercelDeployment,
  type VercelEnvVar,
  type VercelRequestAuth,
} from "./vercel-manager.js";

/**
 * Write-capable Vercel operations, deliberately separate from the read-safe
 * {@link VercelManager} — the same split as `git-manager` / `git-actions` and
 * `db-peek` / `db-write`.
 *
 * Everything here changes what the internet is serving, so none of it is
 * exposed as an MCP tool: these are reached only from the dashboard's own
 * routes, where a human clicked the button. Callers are expected to confirm
 * production-affecting actions (`promote`, `rollback`, `deleteEnv`) before
 * invoking them.
 *
 * Still intentionally excludes the irreversible ones — deleting deployments or
 * projects — which would need their own guarded surface.
 */
export class VercelActions {
  constructor(private readonly auth: VercelRequestAuth) {}

  /**
   * Rebuild an existing deployment. Vercel inherits the original's settings,
   * env, and git ref, minting a new deployment id — it never mutates the one
   * being retried, so this is safe to offer on a failed build.
   *
   * `target` carries the original's environment through: without it a redeploy
   * of a production deployment would come back as a preview.
   */
  async redeploy(deployment: Pick<VercelDeployment, "uid" | "name" | "target">): Promise<{
    uid: string;
    url: string | null;
  }> {
    const created = await this.request<{ id?: string; uid?: string; url: string | null }>(
      "/v13/deployments?forceNew=1",
      {
        method: "POST",
        body: {
          deploymentId: deployment.uid,
          name: deployment.name,
          ...(deployment.target ? { target: deployment.target } : {}),
        },
      },
    );
    return { uid: created.uid ?? created.id ?? "", url: created.url };
  }

  /** Stop an in-flight build. A finished deployment cannot be canceled (Vercel 400s). */
  async cancel(deploymentId: string): Promise<void> {
    await this.request(`/v12/deployments/${encodeURIComponent(deploymentId)}/cancel`, {
      method: "PATCH",
    });
  }

  /**
   * Point the project's production alias at an existing deployment without
   * rebuilding it. This is what "Promote to Production" does in the Vercel
   * dashboard, and it takes effect immediately.
   */
  async promote(projectId: string, deploymentId: string): Promise<void> {
    await this.request(
      `/v10/projects/${encodeURIComponent(projectId)}/promote/${encodeURIComponent(deploymentId)}`,
      { method: "POST" },
    );
  }

  /**
   * Roll production back *to* an earlier deployment. Distinct from `promote`
   * on the API side: it records a rollback (with its reason) rather than a
   * forward promotion, which is what the project's rollback history reads back.
   */
  async rollback(
    projectId: string,
    deploymentId: string,
    description?: string,
  ): Promise<void> {
    const path = `/v1/projects/${encodeURIComponent(projectId)}/rollback/${encodeURIComponent(deploymentId)}`;
    const params = description ? `?description=${encodeURIComponent(description)}` : "";
    await this.request(`${path}${params}`, { method: "POST" });
  }

  /**
   * Add a new variable. `type` defaults to `encrypted` (Vercel's "Sensitive"),
   * which never reads back over the API — `plain` is for values a build script
   * needs to see, not secrets.
   */
  async createEnv(
    projectId: string,
    input: { key: string; value: string; target: string[]; type?: "encrypted" | "plain" },
  ): Promise<VercelEnvVar> {
    const created = await this.request<RawEnvVar | { created?: RawEnvVar[] }>(
      `/v10/projects/${encodeURIComponent(projectId)}/env`,
      {
        method: "POST",
        body: {
          key: input.key,
          value: input.value,
          target: input.target,
          type: input.type ?? "encrypted",
        },
      },
    );
    const raw =
      "created" in created && created.created?.[0] ? created.created[0] : (created as RawEnvVar);
    return normalizeEnvVar(raw);
  }

  /**
   * Change a variable's value and/or the environments it applies to. The key
   * itself is not editable here — Vercel does not support renaming in place,
   * so a rename is a delete-and-recreate the UI does not offer as one step.
   */
  async updateEnv(
    projectId: string,
    envId: string,
    input: { value?: string; target?: string[] },
  ): Promise<VercelEnvVar> {
    const body: Record<string, unknown> = {};
    if (input.value !== undefined) body.value = input.value;
    if (input.target !== undefined) body.target = input.target;
    const raw = await this.request<RawEnvVar>(
      `/v9/projects/${encodeURIComponent(projectId)}/env/${encodeURIComponent(envId)}`,
      { method: "PATCH", body },
    );
    return normalizeEnvVar(raw);
  }

  async deleteEnv(projectId: string, envId: string): Promise<void> {
    await this.request(
      `/v9/projects/${encodeURIComponent(projectId)}/env/${encodeURIComponent(envId)}`,
      { method: "DELETE" },
    );
  }

  private request<T>(path: string, opts: { method: string; body?: unknown }): Promise<T> {
    return vercelRequest<T>(this.auth, path, opts);
  }
}

export const VERCEL_ACTIONS = ["redeploy", "cancel", "promote", "rollback"] as const;
export type VercelActionName = (typeof VERCEL_ACTIONS)[number];

/** Actions that change what production serves — the UI confirms these first. */
export const PRODUCTION_AFFECTING_ACTIONS: ReadonlySet<VercelActionName> = new Set([
  "promote",
  "rollback",
]);

File: crates/nomoreide-core/src/vercel_actions.rs
//! Write-capable Vercel operations, deliberately separate from the read-safe
//! `vercel_manager` — the same split as `git_manager` / git write commands, and
//! the Rust counterpart of `src/core/vercel-actions.ts`.
//!
//! Everything here changes what the internet is serving, so none of it is
//! exposed to an agent: these are reached only from the dashboard's own
//! commands, where a human clicked the button. Callers are expected to confirm
//! production-affecting actions (`promote`, `rollback`) before invoking them.
//!
//! Still intentionally excludes the irreversible ones — deleting deployments or
//! projects — which would need their own guarded surface.

use serde_json::Value;

use super::vercel_manager::{request, request_json, RequestAuth, VercelApiError};
use crate::providers::deploy::{truthy, CreatedDeployment};

pub struct VercelActions {
    auth: RequestAuth,
}

impl VercelActions {
    pub fn new(auth: RequestAuth) -> Self {
        VercelActions { auth }
    }

    /// Rebuild an existing deployment. Vercel inherits the original's settings,
    /// env, and git ref, minting a new deployment id — it never mutates the one
    /// being retried, so this is safe to offer on a failed build.
    ///
    /// `target` carries the original's environment through: without it a
    /// redeploy of a production deployment would come back as a preview.
    /// `name` and `target` travel as the JSON the vendor sent, not as strings:
    /// they are read straight off the original deployment and handed back
    /// unchanged, and narrowing them here would be this layer inventing a type
    /// the vendor never promised.
    pub async fn redeploy(
        &self,
        uid: &str,
        name: &Value,
        target: &Value,
    ) -> Result<CreatedDeployment, VercelApiError> {
        let mut body = serde_json::json!({ "deploymentId": uid, "name": name });
        // Omitted rather than sent as null when the original had none: Vercel
        // reads the key's presence, and a null target is a preview.
        if truthy(target) {
            body["target"] = target.clone();
        }
        let created = request_json(
            &self.auth,
            "POST",
            "/v13/deployments?forceNew=1",
            Some(&body),
        )
        .await?;
        Ok(CreatedDeployment {
            id: created
                .get("uid")
                .or_else(|| created.get("id"))
                .and_then(Value::as_str)
                .unwrap_or_default()
                .to_string(),
            url: created.get("url").cloned().unwrap_or(Value::Null),
        })
    }

    /// Stop an in-flight build. A finished deployment cannot be canceled
    /// (Vercel answers 400).
    pub async fn cancel(&self, deployment_id: &str) -> Result<(), VercelApiError> {
        let path = format!(
            "/v12/deployments/{}/cancel",
            urlencoding::encode(deployment_id)
        );
        request(&self.auth, "PATCH", &path, None).await?;
        Ok(())
    }

    /// Point the project's production alias at an existing deployment without
    /// rebuilding it. This is what "Promote to Production" does in the Vercel
    /// dashboard, and it takes effect immediately.
    pub async fn promote(
        &self,
        project_id: &str,
        deployment_id: &str,
    ) -> Result<(), VercelApiError> {
        let path = format!(
            "/v10/projects/{}/promote/{}",
            urlencoding::encode(project_id),
            urlencoding::encode(deployment_id)
        );
        request(&self.auth, "POST", &path, None).await?;
        Ok(())
    }

    /// Roll production back *to* an earlier deployment. Distinct from
    /// `promote` on the API side: it records a rollback (with its reason)
    /// rather than a forward promotion, which is what the project's rollback
    /// history reads back.
    pub async fn rollback(
        &self,
        project_id: &str,
        deployment_id: &str,
        description: Option<&str>,
    ) -> Result<(), VercelApiError> {
        let mut path = format!(
            "/v1/projects/{}/rollback/{}",
            urlencoding::encode(project_id),
            urlencoding::encode(deployment_id)
        );
        if let Some(description) = description {
            path.push_str(&format!(
                "?description={}",
                urlencoding::encode(description)
            ));
        }
        request(&self.auth, "POST", &path, None).await?;
        Ok(())
    }

    /// Add a variable.
    ///
    /// `encrypted` (Vercel's "Sensitive") is the default, and its value never
    /// reads back over the API afterwards — `plain` is for values a build
    /// script needs to see, not for secrets.
    pub async fn create_env(
        &self,
        project_id: &str,
        key: &str,
        value: &str,
        environments: &[String],
        kind: &str,
    ) -> Result<Value, VercelApiError> {
        let path = format!("/v10/projects/{}/env", urlencoding::encode(project_id));
        let body = serde_json::json!({
            "key": key,
            "value": value,
            "target": environments,
            "type": kind,
        });
        let created = request_json(&self.auth, "POST", &path, Some(&body)).await?;
        // This endpoint answers either the record itself or a `{ created: [] }`
        // batch, depending on how Vercel felt about the request.
        // The record is returned as Vercel sent it; the provider layer is
        // where a vendor record becomes a neutral one, and doing it here would
        // be a second place that mapping lives.
        Ok(created
            .get("created")
            .and_then(Value::as_array)
            .and_then(|batch| batch.first())
            .cloned()
            .unwrap_or(created))
    }

    /// Change a variable's value and/or the environments it applies to.
    ///
    /// The key itself is not editable: Vercel does not support renaming in
    /// place, so a rename is a delete-and-recreate the UI does not offer as one
    /// step.
    pub async fn update_env(
        &self,
        project_id: &str,
        env_id: &str,
        value: Option<&str>,
        environments: Option<&[String]>,
    ) -> Result<Value, VercelApiError> {
        let path = format!(
            "/v9/projects/{}/env/{}",
            urlencoding::encode(project_id),
            urlencoding::encode(env_id)
        );
        // Only the fields the caller named: an absent `value` means "leave it",
        // which is not the same as setting it to the empty string.
        let mut body = serde_json::Map::new();
        if let Some(value) = value {
            body.insert("value".into(), Value::String(value.to_string()));
        }
        if let Some(environments) = environments {
            body.insert("target".into(), serde_json::json!(environments));
        }
        request_json(&self.auth, "PATCH", &path, Some(&Value::Object(body))).await
    }

    pub async fn delete_env(&self, project_id: &str, env_id: &str) -> Result<(), VercelApiError> {
        let path = format!(
            "/v9/projects/{}/env/{}",
            urlencoding::encode(project_id),
            urlencoding::encode(env_id)
        );
        request(&self.auth, "DELETE", &path, None).await?;
        Ok(())
    }
}
